desc/particles.py

```diff
@@ -0,0 +1,1239 @@
+
+        Returns
+        -------
+        x0 : jax.Array, shape(N,D)
+            Initial particle positions and velocities, where D is the dimensionality of
+            the trajectory model, which includes 3D spatial dimensions and depending on
+            the model, parallel velocity and total velocity. The initial positions are
+            in the frame of the model.
+        args : jax.Array, shape(N,M)
+            Additional arguments needed by the model, such as mass, charge, and
+            magnetic moment (mv⊥²/2|B|) of each particle. M is the number of arguments
+            requested by the model which is equal to len(model.args). N is the number
+            of particles.
+        """
+        x = jnp.array([self.R0, self.phi0, self.Z0]).T
+        params = field.params_dict
+        if model.frame == "flux":
+            if not isinstance(field, Equilibrium):
+                raise ValueError(
+                    "Mapping from lab to flux coordinates requires an Equilibrium. "
+                    "Please use Equilibrium object with the model."
+                )
+            # there is not much we can do for the guess here, but we need the
+            # guess for jit purposes
+            x_guess = jnp.zeros(x.shape)
+            x_guess = x_guess.at[:, 2].set(self.phi0)
+            x_guess = x_guess.at[:, 1].set(jnp.arctan2(self.Z0, self.R0))
+            tol = 1e-8
+            x, out = field.map_coordinates(
+                coords=x,
+                inbasis=("R", "phi", "Z"),
+                outbasis=("rho", "theta", "zeta"),
+                maxiter=200,
+                guess=x_guess,
+                tol=tol,
+                full_output=True,
+            )
+            x = eqx.error_if(
+                x,
+                (out[0] > tol).any(),
+                "Mapping from lab to flux coordinates failed to achieve tolerance "
+                f"{tol:.2e}. Make sure the points lie in the equilibrium.",
+            )
+            if field.iota is None:
+                iota = field.get_profile("iota")
+                params["i_l"] = iota.params
+                kwargs["iota"] = iota
+        elif model.frame == "lab":
+            if isinstance(field, Equilibrium):
+                raise NotImplementedError(
+                    "If you have an Equilibrium object, you should use the model "
+                    "in flux frame. Since trying to integrate in lab frame will "
+                    "require multiple coordinate mappings, it is not implemented."
+                )
+            x = x
+        else:
+            raise NotImplementedError
+
+        return super()._return_particles(
+            x=x,
+            v=self.v0,
+            vpar=self.vpar0,
+            model=model,
+            field=field,
+            params=params,
+            **kwargs,
+        )
+
+
+class CurveParticleInitializer(AbstractParticleInitializer):
+    """Randomly sample particles starting on a curve.
+
```

Comment:
explain the sampling more, it is specifically sampling the positions randomly from the curve on the given discrete grid, with higher probability given to points with larger surface area. Then each particle additioally has its xi assigned randomly